### PR TITLE
feat(ec2): add tag-based filtering to EC2 explorer node

### PR DESCRIPTION
**Pull request by shivkumr**
**Problem**
When working with many EC2 instances, the AWS Explorer EC2 tree can be overwhelming. There's no way to narrow down the list to instances matching a specific tag.

**Solution**
Adds a "Filter by Tag" action to the EC2 parent node in the AWS Explorer. Users can click the filter icon (funnel) on the EC2 node row, enter a `key=value` tag pair, and the tree reloads showing only matching instances.

**Changes**

- `Ec2ParentNode`: Added `tagFilter` state, `setTagFilter(key, value)` method, and label updates `(EC2 [tag: key=value])`
- `activation.ts: `Registered `aws.ec2.filterByTag` command with input box and parsing logic
- `{} package.json `: Declared command and inline menu contribution for awsEc2ParentNode
- Passes filter to existing `getInstances(filters?) ` - no EC2 client changes needed
 
**Behavior**

- Click filter icon → input box with key=value (e.g. Env=prod) placeholder
- Submit key=value → tree filters to matching instances, label shows EC2 [tag: Env=prod]
- Submit empty → clears filter, label resets to EC2
- Cancel (Escape) → no-op, filter unchanged
- Handles edge cases: multiple = in value, no =, whitespace-only input

**Testing**

- 7 unit tests covering filter state, label updates, API passthrough, and input parsing edge cases
- 4 property-based tests using fast-check (100 iterations each) validating filter construction, clearing, input parsing, and whitespace handling

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be squash-merged at release time.
- Your code changes must meet the guidelines in [CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/awsService/ec2/explorer/ec2ParentNode.ts [hunk 1]
```diff
@@ -27,6 +29,17 @@ export class Ec2ParentNode extends AWSTreeNodeBase {
     ) {
         super('EC2', vscode.TreeItemCollapsibleState.Collapsed)
         this.ec2InstanceNodes = new Map<string, Ec2InstanceNode>()
+        this.tagFilter = undefined
+    }
+
+    public setTagFilter(key: string, value: string): void {
+        if (key !== '' && value !== '') {
+            this.tagFilter = [{ Name: `tag:${key}`, Values: [value] }]
+            this.label = `EC2 [tag: ${key}=${value}]`
```
Comment (yuriivv) [inline packages/core/src/awsService/ec2/explorer/ec2ParentNode.ts]: EC2 tags can have empty values (e.g. tags like `prod-fleet` and `gamma-fleet` vs `fleet=prod` and `fleet=gamma`). `setTagFilter` requires both `key !== ''` AND `value !== ''`, so `Env=` (valid filter) is silently treated as "clear filter." If a user types just `Env` (no `=`), the handler sets `key='Env', value=''`, which also clears the filter with no feedback. I think that it should work like this:
- `Env=prod` → filters instances with `Env` tag having value `prod`
- `Prod` → filters instances with `Prod` tag (no value)

File: packages/core/src/awsService/ec2/activation.ts [hunk 2]
```diff
@@ -78,6 +78,36 @@ export async function activate(ctx: ExtContext): Promise<void> {
...
+            if (eqIndex === -1) {
+                key = trimmed
+                value = ''
+            } else {
+                key = trimmed.substring(0, eqIndex)
+                value = trimmed.substring(eqIndex + 1)
+            }
+
+            node.setTagFilter(key, value)
+            await node.refreshNode()
```
Comment (yuriivv) [inline packages/core/src/awsService/ec2/activation.ts]: Can you please move this into commands.ts to keep `activate` function neat?

File: packages/toolkit/package.json [hunk 3]
```diff
@@ -1929,6 +1929,11 @@
                     "when": "view == aws.explorer && viewItem =~ /^(awsEc2(Running|Stopped|Pending)Node)$/",
                     "group": "2@0"
                 },
+                {
+                    "command": "aws.ec2.filterByTag",
+                    "when": "viewItem =~ /^(awsEc2ParentNode)$/",
```
Comment (yuriivv) [inline packages/toolkit/package.json]: Missing `view == aws.explorer &&` prefix that all other EC2 menu items use. Could potentially match nodes in other views with the same context value.

File: packages/toolkit/package.json [hunk 4]
```diff
@@ -3406,6 +3411,18 @@
                     }
                 }
             },
+            {
+                "command": "aws.ec2.filterByTag",
+                "title": "Filter by Tag",
```
Comment (yuriivv) [inline packages/toolkit/package.json]: This should use `%AWS.command.ec2.XXX%` reference resolved from packages/core/package.nls.json:
- `"title": "%AWS.command.ec2.filterByTag%"` in packages/toolkit/package.json
- `"AWS.command.ec2.filterByTag": "Filter by Tag"` added to packages/core/package.nls.json

File: packages/core/package.json [hunk 5]
```diff
@@ -558,6 +558,7 @@
         "circular-dependency-plugin": "^5.2.2",
         "css-loader": "^6.10.0",
         "esbuild-loader": "2.20.0",
+        "fast-check": "^3.22.0",
```
Comment (yuriivv) [inline packages/core/package.json]: `fast-check` is added to be used only in one test file. The project uses mocha + sinon + assert. The 4 property tests can be rewritten as parameterized mocha tests with explicit test vectors.

**Comment by yuriivv** [comment]
Please add change entry for this user-facing feature. See https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#changelog

File: packages/core/package.json (hunk 5 above)
Comment (yuriivv) [reply packages/core/package.json]: Consider something like this (haven't tested)
```
import assert from 'assert'
import { Ec2ParentNode } from '../../../../awsService/ec2/explorer/ec2ParentNode'
import { Ec2Client } from '../../../../shared/clients/ec2'

describe('ec2ParentNode property tests', function () {
    const testRegion = 'us-east-1'
    const testPartition = 'aws'

    function createNode(): Ec2ParentNode {
        const client = new Ec2Client(testRegion)
        return new Ec2ParentNode(testRegion, testPartition, client)
    }

    const tagCases = [
        ['Env', 'prod'],
        ['Name', 'my-server'],
        ['team', 'backend'],
        ['key-with-dashes', 'value_with_underscores'],
        ['CaseSensitive', 'MiXeD'],
        ['unicode', '日本語'],
        ['a', 'b'],
    ] as const

    for (const [key, value] of tagCases) {
        it(`setTagFilter produces correct filter and label for ${key}=${value}`, function () {
            const node = createNode()
            node.setTagFilter(key, value)

            const filter = (node as any).tagFilter
            assert.deepStrictEqual(filter, [{ Name: `tag:${key}`, Values: [value] }])
            assert.strictEqual(node.label, `EC2 [tag: ${key}=${value}]`)
        })
    }

    for (const [key, value] of tagCases) {
        it(`clearing filter resets state after setting ${key}=${value}`, function () {
            const node = createNode()
            node.setTagFilter(key, value)
            assert.notStrictEqual((node as any).tagFilter, undefined)

            node.setTagFilter('', '')
            assert.strictEqual((node as any).tagFilter, undefined)
            assert.strictEqual(node.label, 'EC2')
        })
    }

    const splitCases = [
        ['Env=prod', 'Env', 'prod'],
        ['Name=my=server', 'Name', 'my=server'],
        ['key=a=b=c', 'key', 'a=b=c'],
        ['tag=', 'tag', ''],
        ['x=y', 'x', 'y'],
    ] as const

    for (const [input, expectedKey, expectedValue] of splitCases) {
        it(`input "${input}" splits into key="${expectedKey}" value="${expectedValue}"`, function () {
            const trimmed = input.trim()
            const eqIndex = trimmed.indexOf('=')
            const key = trimmed.substring(0, eqIndex)
            const value = trimmed.substring(eqIndex + 1)

            assert.strictEqual(key, expectedKey)
            assert.strictEqual(value, expectedValue)
            assert.ok(!key.includes('='))
        })
    }

    const whitespaceCases = [' ', '  ', '\t', '\n', ' \t\n ', '\r\n']

    for (const ws of whitespaceCases) {
        it(`whitespace-only input ${JSON.stringify(ws)} clears filter`, function () {
            const trimmed = ws.trim()
            assert.strictEqual(trimmed, '')

            const node = createNode()
            node.setTagFilter('SomeKey', 'SomeValue')
            node.setTagFilter('', '')

            assert.strictEqual((node as any).tagFilter, undefined)
            assert.strictEqual(node.label, 'EC2')
        })
    }
})
```

File: .changes/next-release/Feature-31d7baff-32a9-465d-b761-7b3664d22182.json
Comment (yuriivv) [inline .changes/next-release/Feature-31d7baff-32a9-465d-b761-7b3664d22182.json:1]: Did you use `npm run newChange -w packages/toolkit` from CONTRIBUTING.md to generate this? It normally generates a json file with a UUID in its name. Overall it looks correct, but I'm not sure if there's any automation in the release process that relies on that ID.

**Comment by shivkumr** [comment]
> Please add change entry for this user-facing feature. See https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#changelog

I initially created it manually. Regenerated it using npm run newChange -w packages/toolkit as recommended. Updated in the latest push.

File: packages/core/src/awsService/ec2/explorer/ec2ParentNode.ts (hunk 1 above)
Comment (shivkumr) [reply packages/core/src/awsService/ec2/explorer/ec2ParentNode.ts]: Fixed — setTagFilter now only requires a non-empty key. Empty values are valid.

File: packages/core/src/awsService/ec2/activation.ts (hunk 2 above)
Comment (shivkumr) [reply packages/core/src/awsService/ec2/activation.ts]: Done — moved filterByTag to commands.ts.

File: packages/toolkit/package.json (hunk 3 above)
Comment (shivkumr) [reply packages/toolkit/package.json]: Fixed — added the view scope prefix.

File: packages/toolkit/package.json (hunk 4 above)
Comment (shivkumr) [reply packages/toolkit/package.json]: Done — using %AWS.command.ec2.filterByTag% now, added to package.nls.json.

File: packages/core/package.json (hunk 5 above)
Comment (shivkumr) [reply packages/core/package.json]: Removed and rewrote as parameterized mocha tests with explicit test vectors.

File: packages/core/src/awsService/ec2/commands.ts [hunk 6]
```diff
@@ -59,3 +60,33 @@ export async function copyInstanceId(instanceId: string): Promise<void> {
 export async function openLogDocument(node?: Ec2InstanceNode): Promise<void> {
     return await showFile(ec2LogSchema.form(await getSelection(node)))
 }
+
+export async function filterByTag(node?: Ec2ParentNode): Promise<void> {
+    if (!node) {
+        return
+    }
+
+    const input = await vscode.window.showInputBox({
```
Comment (yuriivv) [inline packages/core/src/awsService/ec2/commands.ts:69]: nit: Other filter UIs in the codebase (e.g., CloudWatch Logs) use the shared showInputBox from ../../shared/ui/inputPrompter for consistent styling and validation. Not a blocker, leaving this for future reference

File: packages/core/src/awsService/ec2/explorer/ec2ParentNode.ts [hunk 7]
```diff
@@ -27,6 +29,17 @@ export class Ec2ParentNode extends AWSTreeNodeBase {
     ) {
         super('EC2', vscode.TreeItemCollapsibleState.Collapsed)
         this.ec2InstanceNodes = new Map<string, Ec2InstanceNode>()
+        this.tagFilter = undefined
+    }
+
+    public setTagFilter(key: string, value: string): void {
+        if (key !== '') {
+            this.tagFilter = [{ Name: `tag:${key}`, Values: [value] }]
+            this.label = value !== '' ? `EC2 [tag: ${key}=${value}]` : `EC2 [tag: ${key}]`
```
Comment (yuriivv) [inline packages/core/src/awsService/ec2/explorer/ec2ParentNode.ts]: Key-only filtering (e.g. typing just `PROD`) doesn't actually work — `{ Name: 'tag:PROD', Values: [''] }` asks EC2 for tags with an empty string value, not "tag exists." You need the `tag-key` filter for that.

Something like:

```
    public setTagFilter(key: string, value: string): void {
        if (key !== '') {
            if (value !== '') {
                this.tagFilter = [{ Name: `tag:${key}`, Values: [value] }]
                this.label = `EC2 [tag: ${key}=${value}]`
            } else {
                this.tagFilter = [{ Name: 'tag-key', Values: [key] }]
                this.label = `EC2 [tag: ${key}]`
            }
        } else {
            this.tagFilter = undefined
            this.label = 'EC2'
        }
    }
```

Comment (shivkumr) [reply packages/core/src/awsService/ec2/explorer/ec2ParentNode.ts]: Fixed - key-only input now uses { Name: 'tag-key', Values: [key] } to filter by tag existence rather than matching an empty value.

**Review by yuriivv — APPROVED** [review]
